bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -428,8 +443,29 @@ private static boolean isConvertibleToArrayType(Object sourceValue, BArrayType t
         }
         ArrayValue source = (ArrayValue) sourceValue;
         Type targetTypeElementType = targetType.getElementType();
+        if (source.getType().getTag() == TypeTags.ARRAY_TAG) {
+            Type sourceElementType = ((BArrayType) source.getType()).getElementType();
+            if (isNumericType(sourceElementType) && isNumericType(targetTypeElementType)) {
+                return true;
+            }
+        }
+        Set<Type> convertibleTypes;
         for (int i = 0; i < source.size(); i++) {
-            if (getConvertibleTypes(source.get(i), targetTypeElementType, unresolvedValues).size() != 1) {
+            convertibleTypes = getConvertibleTypes(source.get(i), targetTypeElementType, unresolvedValues);
+            if (convertibleTypes.size() != 1 && !convertibleTypes.contains(TypeChecker.getType(source.get(i)))
+                    && !hasIntegerSubTypes(convertibleTypes)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean hasIntegerSubTypes(Set<Type> convertibleTypes) {
+        if (convertibleTypes.size() < 2) {
```

Comment:
Do we need this check?

Reply:
convertibleTypes.size() != 1  and empty is already checked in previous conditions.

Reply:
Yeah, `convertibleTypes.size() < 2` is kind of repeated check, that is not needed here. But we need to preserve the `convertibleTypes.size()  == 0` check for other places. Moved that check to outside of this method. 
Addressed with https://github.com/ballerina-platform/ballerina-lang/pull/30710/commits/688ab907b7244cd39e24c251849c329e7f7fbe24